[Bug]: SA (Suffix Automaton) speculative decoding crashes the executor event loop when a prompt is near `max_seq_len` — `default_max_tokens` is deduced against the draft-inflated `max_seq_len`

### System Info

- TensorRT-LLM: PyTorch backend, `trtllm-serve` (OpenAI chat completions)
- Speculative decoding: `decoding_type: "SA"` (Suffix Automaton, one-model), overlap scheduler enabled
- `max_seq_len = 9600`, `max_draft_len = 4`

### Who can help?

Speculative decoding / SA framework owners (feature introduced in #11434).


### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction


1. Serve a model with `max_seq_len: 9600` and SA speculative decoding enabled (overlap scheduler on):
   ```yaml
   max_seq_len: 9600
   speculative_config:
     decoding_type: SA
     max_draft_len: 4
   ```
2. Send a chat completion whose prompt is **9608 tokens** (i.e. slightly *above* the configured `max_seq_len` of 9600, but below the internally-inflated cap of 9611) with `max_tokens: 512`.




### Expected behavior

The request should be rejected with a clean per-request error (e.g. HTTP 400 / `RequestError`) because the prompt (9608) exceeds the configured `max_seq_len` (9600). Other in-flight requests should be unaffected and the server should stay healthy.

### actual behavior

The request is **admitted**, then overflows the KV cache during a speculative generation step. This triggers **two separate failures** — the second of which is fatal to the whole server:

1. The offending request gets a clean `RequestError` and returns `400 Bad Request`. On its own this would be harmless.
2. But the same exception also propagates through `_forward_step` into the executor loop, where `_executor_loop_overlap` hits `assert sample_state is not None, "Sampling failed"` (`py_executor.py:2941`). This `AssertionError` escapes `event_loop()`, `_event_loop_wrapper` re-raises it (`py_executor.py:723-726`), and **the worker thread (`Thread-4`) dies.** Its `finally` runs `_executor_loop_cleanup()` (`py_executor.py:1543-1554`), setting `is_shutdown = True`.

**Resulting state — a "zombie" server (worse than a clean crash):**
- The worker thread is the only thing that pulls from the request queue and runs model forwards. Once dead, **no request is ever processed again.**
- `is_shutdown = True` makes `_await_any_response` (`py_executor.py:4365-4366`) wake all waiters immediately, so every in-flight and subsequent request errors out / returns empty.
- The worker thread is `daemon=True` (`py_executor.py:770`), so **the OS process does not exit** — uvicorn/FastAPI keeps listening on the port. Health/liveness probes may still report "up" while all inference is permanently broken, so orchestrators (k8s, etc.) may not restart it automatically. A manual restart is required.

So a single over-length prompt causes a **full inference outage for all clients**, not an isolated per-request error.

```
[W] [executor] User-specified `max_tokens` (512) is greater than deduced `default_max_tokens` (3), using default_max_tokens instead.
...
  File ".../_torch/attention_backend/trtllm.py", line 494, in prepare
    assert self.kv_lens[:self.num_seqs].max(
AssertionError: The max KV cache length of input sequences (9620) exceeds the KV cache manager's maximum supported length (9611).
[E] [_torch] Encountered an error in forward function: The max KV cache length of input sequences (9620) exceeds the KV cache manager's maximum supported length (9611).
tensorrt_llm.executor.utils.RequestError: The max KV cache length of input sequences (9620) exceeds the KV cache manager's maximum supported length (9611).
INFO:     127.0.0.1:57804 - "POST /v1/chat/completions HTTP/1.1" 400 Bad Request
...
  File ".../_torch/pyexecutor/py_executor.py", line 2941, in _executor_loop_overlap
    assert sample_state is not None, "Sampling failed"
AssertionError: Sampling failed
Exception in thread Thread-4 (_event_loop_wrapper):
```


### additional notes

```python
# tensorrt_llm/_torch/pyexecutor/py_executor_creator.py:599-613
model_engine_max_seq_len = model_engine.max_seq_len   # net = 9600
net_max_seq_len = model_engine_max_seq_len
if not llm_args.disable_overlap_scheduler and spec_config is not None:
    model_engine_max_seq_len += spec_config.tokens_per_gen_step - 1        # +D   (=4)
if spec_config is not None:
    model_engine_max_seq_len += get_num_extra_kv_tokens(spec_config)       # +D-1 (=3, SA is one-engine)
    model_engine_max_seq_len += spec_config.tokens_per_gen_step - 1        # +D   (=4)
max_seq_len = model_engine_max_seq_len                                     # = 9600 + 3D - 1 = 9611
```

For SA: `tokens_per_gen_step = 1 + max_total_draft_tokens = 1 + D`, and `get_num_extra_kv_tokens = max_draft_len - 1` because `SpeculativeDecodingMode.is_sa()` → `use_one_engine() == True`. With `D = 4`: inflated `max_seq_len = 9600 + 11 = 9611`, which matches `kv_cache_manager.max_seq_len` in the crash.

The problem: this **inflated** value (9611) is propagated to `PyExecutor.max_seq_len` and then to the worker used for request admission / token deduction:

- `py_executor.py:623` → `self.max_seq_len = max_seq_len` (inflated)
- `base_worker.py:246` → `self.max_seq_len = _executor.max_seq_len` (inflated)
- `base_worker.py:520` → `default_max_tokens = max_seq_len - splited_prompt_len - query_token_len`

So `default_max_tokens = 9611 - 9608 = 3 > 0`, and the request is admitted even though the prompt (9608) is larger than the configured net `max_seq_len` (9600). The `prompt_len > max_seq_len` guard at `base_worker.py:516-518` only runs on the legacy `executor_config` path, not the `llm_args` path, so nothing rejects it.

Result: the draft-token reserve (`3D-1`) is being handed out as usable **golden** context budget. Golden generation is allowed up to 9611, and the SA step then writes `golden + draft(D) + extra_kv(D-1)` on top → **9620 > 9611** → assertion → event loop dies.

Net effect: **an over-length prompt (which should be a graceful 400) instead crashes the entire executor thread**, matching the concern in #6329.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and checked the [documentation](https://nvidia.github.io/TensorRT-LLM/) and [examples](https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples) for answers to frequently asked questions.

--- a/tensorrt_llm/_torch/pyexecutor/py_executor_creator.py
+++ b/tensorrt_llm/_torch/pyexecutor/py_executor_creator.py
@@ -1009,7 +1009,7 @@ def create_py_executor(
             kv_connector_manager=kv_connector_manager
             if not estimating_kv_cache else None,
             resource_governor_queue=resource_governor_queue,
-            max_seq_len=max_seq_len,
+            max_seq_len=net_max_seq_len,
             max_batch_size=max_batch_size,
             max_beam_width=max_beam_width,
             max_num_tokens=max_num_tokens,
@@ -1084,7 +1084,7 @@ def create_py_executor(
                 garbage_collection_gen0_threshold,
                 kv_connector_manager=kv_connector_manager,
                 resource_governor_queue=resource_governor_queue,
-                max_seq_len=max_seq_len,
+                max_seq_len=net_max_seq_len,
                 max_batch_size=max_batch_size,
                 max_beam_width=max_beam_width,
                 max_num_tokens=max_num_tokens,

--- a/tests/unittest/llmapi/test_llm.py
+++ b/tests/unittest/llmapi/test_llm.py
@@ -17,7 +17,8 @@ from tensorrt_llm import LLM
 from tensorrt_llm.bindings import executor as tllm
 from tensorrt_llm.executor import GenerationResultBase, RequestError
 from tensorrt_llm.llmapi import (KvCacheConfig, KvCacheRetentionConfig,
-                                 LookaheadDecodingConfig, RequestOutput)
+                                 LookaheadDecodingConfig, RequestOutput,
+                                 SADecodingConfig)
 from tensorrt_llm.llmapi.llm_args import DynamicBatchConfig, SchedulerConfig
 from tensorrt_llm.llmapi.llm_utils import _ParallelConfig
 from tensorrt_llm.llmapi.tokenizer import (TokenizerBase, TransformersTokenizer,
@@ -1104,6 +1105,38 @@ def _test_llm_capture_request_error(pytorch_backend: bool, tp_size: int = 1):
     # Both backends now consistently raise RequestError for max_num_tokens validation
     with pytest.raises(RequestError):
         llm.generate(prompt)
+
+
+def test_oversized_prompt_rejected_with_inflated_max_seq_len():
+    """Oversized prompt is rejected gracefully when max_seq_len is inflated internally.
+
+    With speculative decoding, KV cache max_seq_len is inflated to reserve space
+    for draft tokens. Admission control must use the net (user-configured) limit,
+    not the inflated value. Otherwise oversized prompts crash the event loop.
+    """
+    max_seq_len = 256
+
+    llm = LLM(
+        model=llama_model_path,
+        max_seq_len=max_seq_len,
+        max_batch_size=1,
+        max_num_tokens=max_seq_len,
+        kv_cache_config=KvCacheConfig(enable_block_reuse=False),
+        speculative_config=SADecodingConfig(max_draft_len=4),
+        disable_overlap_scheduler=True,
+    )
+
+    # Prompt that fills the entire net budget -> 0 tokens left for output -> rejected
+    oversized_prompt_ids = [1] * (max_seq_len + 1)
+    with pytest.raises(RequestError):
+        llm.generate([oversized_prompt_ids],
+                     sampling_params=SamplingParams(max_tokens=1))
+
+    # Executor must still be alive
+    result = llm.generate(["Hello"],
+                          sampling_params=SamplingParams(max_tokens=5))
+    assert len(result) == 1
+    assert len(result[0].outputs[0].text) > 0
 
 
 def test_llm_shutdown_executor():

--- a/tests/unittest/llmapi/test_llm.py
+++ b/tests/unittest/llmapi/test_llm.py
@@ -1120,7 +1120,7 @@ def test_oversized_prompt_rejected_with_inflated_max_seq_len():
         model=llama_model_path,
         max_seq_len=max_seq_len,
         max_batch_size=1,
-        max_num_tokens=max_seq_len,
+        max_num_tokens=(max_seq_len + 32),
         kv_cache_config=KvCacheConfig(enable_block_reuse=False),
         speculative_config=SADecodingConfig(max_draft_len=4),
         disable_overlap_scheduler=True,

